### ISSUE TITLE
SmimeUtil configurable signature algorithm

### ISSUE THREAD

**Issue by kkrupka** [bbottema/simple-java-mail#478]
I'm not sure whether this issue belongs to this project or to [java-utils-mail-smime](https://github.com/simple-java-mail/java-utils-mail-smime) or whether I have overseen anything.

Recently, I have tried to incorporate a new S-MIME certificate into my property-based configuration. With the old provider the S-MIME certificate it worked smoothly. Consequently, I replaced keystore name, password, alias and alias password to the new certificate. When trying to send a signed email I received a SmimeException with underlying NPE

```
Caused by: java.lang.NullPointerException
	at org.simplejavamail.utils.mail.smime.SmimeUtil.getCertificateStore(SmimeUtil.java:411) ~[utils-mail-smime-2.1.2.jar:?]
	at org.simplejavamail.utils.mail.smime.SmimeUtil.getGenerator(SmimeUtil.java:371) ~[utils-mail-smime-2.1.2.jar:?]
	at org.simplejavamail.utils.mail.smime.SmimeUtil.sign(SmimeUtil.java:356) ~[utils-mail-smime-2.1.2.jar:?]
	at org.simplejavamail.utils.mail.smime.SmimeUtil.sign(SmimeUtil.java:457) ~[utils-mail-smime-2.1.2.jar:?]
	at org.simplejavamail.utils.mail.smime.SmimeUtil.sign(SmimeUtil.java:451) ~[utils-mail-smime-2.1.2.jar:?]
	at org.simplejavamail.utils.mail.smime.SmimeUtil.sign(SmimeUtil.java:434) ~[utils-mail-smime-2.1.2.jar:?]
	at org.simplejavamail.internal.smimesupport.SMIMESupport.signMessageWithSmime(SMIMESupport.java:458) ~[smime-module-8.1.3.jar:?]
	at org.simplejavamail.converter.internal.mimemessage.SpecializedMimeMessageProducer.populateMimeMessage(SpecializedMimeMessageProducer.java:81) ~[simple-java-mail-8.1.3.jar:?]
	at org.simplejavamail.converter.internal.mimemessage.MimeMessageProducerHelper.produceMimeMessage(MimeMessageProducerHelper.java:58) ~[simple-java-mail-8.1.3.jar:?]
	at org.simplejavamail.mailer.internal.SessionBasedEmailToMimeMessageConverter.convertMimeMessage(SessionBasedEmailToMimeMessageConverter.java:127) ~[simple-java-mail-8.1.3.jar:?]
	at org.simplejavamail.mailer.internal.SessionBasedEmailToMimeMessageConverter.convertAndLogMimeMessage(SessionBasedEmailToMimeMessageConverter.java:102) ~[simple-java-mail-8.1.3.jar:?]
	at org.simplejavamail.mailer.internal.SessionBasedEmailToMimeMessageConverter.convertAndLogMimeMessage(SessionBasedEmailToMimeMessageConverter.java:79) ~[simple-java-mail-8.1.3.jar:?]
	at org.simplejavamail.mailer.internal.util.TransportRunner.lambda$sendMessage$0(TransportRunner.java:52) ~[simple-java-mail-8.1.3.jar:?]
	at org.simplejavamail.mailer.internal.util.TransportRunner.sendUsingConnectionPool(TransportRunner.java:84) ~[simple-java-mail-8.1.3.jar:?]
	at org.simplejavamail.mailer.internal.util.TransportRunner.runOnSessionTransport(TransportRunner.java:69) ~[simple-java-mail-8.1.3.jar:?]
	at org.simplejavamail.mailer.internal.util.TransportRunner.sendMessage(TransportRunner.java:51) ~[simple-java-mail-8.1.3.jar:?]
	at org.simplejavamail.mailer.internal.SendMailClosure.executeClosure(SendMailClosure.java:69) ~[simple-java-mail-8.1.3.jar:?]
	... 76 more

```

Consequently, I looked at the SmimeUtil class and found our that the signature algorithm is set to SHA256withRSA by default. Comparing my old S-MIME certificate with the new one inspecting it with `keytool -v -list -keystore /path/to/keystore.p12` I have found out that the old signature algorithm was SHA256withRSA, but for the new one it is SHA386withRSA.

**Questions**
1. Generally, could that be a reason for the exception at all or have I been on the wrong track?
2. If 1. applies: Have I overseen a configuration opportunity in simplejavamail.properties or is it possible, that this can be added to the properties configuration (or anywhere else).

**Comment by bbottema** [comment]
Hhm, it's a bit difficult to trace back, but it seems the key is empty from the moment it is attempted to be obtained from the PKCS12 store. Is it possible you used the wrong name as key alias? I noticed Simple Java Mail doesn't do a null-check there, so that's something to be improved.

**Comment by kkrupka** [comment]
You are right. The reason was the alias. However, it wasn't wrong by me, it looks like I have received it by Geant wrong or missing. Printin the information with keytool told that the alias is "1", but that wasn't working. So, I have changed the alias with `keytool -changealias -alias "1" -destalias "new-alias" -keystore /path/to/keystore` and now it works.

Thanks so much. I have tested for days and contacted our responsible department until you wrote, The weird thing is, I tried it before renaming the alias, but it wasn't worked. I think my fault was, that I haven't shutdown the instance correctly before deploying the new one and so it probably used an old state.
